### PR TITLE
Clean up Future/Promises API

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
The generics for the existing futures, promises, and listeners are too complicated.
This complication comes from the existence of `ChannelPromise` and `ChannelFuture`, which forces listeners to care about the particular _type_ of future being listened on.

Modification:
* Add a `FutureContextListener` which can take a context object as an additional argument. This allows our listeners to have the channel piped through to them, so they don't need to rely on the `ChannelFuture.channel()` method.
* Make the `FutureListener`, along with the `FutureContextListener` sibling, the default listener API, retiring the `GenericFutureListener` since we no longer need to abstract over the type of the future.
* Change all uses of `ChannelPromise` to `Promise<Void>`.
* Change all uses of `ChannelFuture` to `Future<Void>`.
* Change all uses of `GenericFutureListener` to either `FutureListener` or `FutureContextListener` as needed.
* Remove `ChannelFutureListener` and `GenericFutureListener`.
* Introduce a `ChannelFutureListeners` enum to house the constants that previously lived in `ChannelFutureListener`. These constants now implement `FutureContextListener` and take the `Channel` as a context.
* Remove `ChannelPromise` and `ChannelFuture` — all usages now rely on the plain `Future` and `Promise` APIs.
* Add static factory methods to `DefaultPromise` that allow us to create promises that are initialised as successful or failed.
* Remove `CompleteFuture`, `SucceededFuture`, `FailedFuture`, `CompleteChannelFuture`, `SucceededChannelFuture`, and `FailedChannelFuture`.
* Remove `ChannelPromiseNotifier`.

Result:
Cleaner generics and hopefully more straight forward code.

### CODE REVIEW

**Comment by chrisvest** [comment]
Related issues are #8516 and #8515

Ping @ejona86 @normanmaurer @NiteshKant @njhill

**Comment by hyperxpro** [comment]
IMHO, it looks more like refactored code.

File: common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java [hunk 1]
```diff
@@ -15,53 +15,75 @@
...
-        size = 2;
+    DefaultFutureListeners() {
+        listeners = new Object[4];
     }
 
-    public void add(GenericFutureListener<? extends Future<?>> l) {
-        GenericFutureListener<? extends Future<?>>[] listeners = this.listeners;
-        final int size = this.size;
-        if (size == listeners.length) {
+    public void add(EventListener listener, Object context) {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java:31]: We might consider requiring `context` to be non-null for `FutureContextListener`, so we can reliably determine which of the `addListener()` methods were called by whether context is null within `listeners`. _Technically_ an object could be both a `FutureContextListener` and a `FutureListener`.

Or probably better, could have a `Object NULL_CONTEXT = new Object()` that we use for `context` if it happens to be `null`, and then use `null` to imply `FutureListener`.

File: common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java [hunk 2]
```diff
@@ -15,53 +15,75 @@
...
+        Object[] listeners = this.listeners;
+        int size = this.size;
+        if (size == listeners.length || listener instanceof FutureContextListener && size == listeners.length - 1) {
             this.listeners = listeners = Arrays.copyOf(listeners, size << 1);
         }
-        listeners[size] = l;
-        this.size = size + 1;
+        listeners[size++] = listener;
+        if (listener instanceof FutureContextListener) {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java:38]: I'd much rather just "waste" the memory to always have the context position than have the array context-sensitive and making assumptions about its input.

File: common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java [hunk 3]
```diff
@@ -15,53 +15,75 @@
...
+        }
+        this.size = size;
     }
 
-    public void remove(GenericFutureListener<? extends Future<?>> l) {
-        final GenericFutureListener<? extends Future<?>>[] listeners = this.listeners;
+    public void remove(EventListener listener) {
+        final Object[] listeners = this.listeners;
         int size = this.size;
         for (int i = 0; i < size; i ++) {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java:47]: This isn't right, as we need to double-step if the current entry is a `FutureContextListener` as we don't want to compare against a Context.

File: common/src/main/java/io/netty/util/concurrent/Promise.java [hunk 4]
```diff
@@ -19,63 +19,61 @@
...
+     * Create a new unfulfilled promise.
      *
+     * @param executor The {@link EventExecutor} the promise should be bound to. Completion callbacks will run on this
+     *                 event executor. It is assumed this executor will protect against {@link StackOverflowError}
+     *                 exceptions. The executor may be used to avoid {@link StackOverflowError} by executing a {@link
+     *                 Runnable} if the stack depth exceeds a threshold.
+     * @param <T>      The result type of the promise.
+     * @return The new promise.
+     */
+    static <T> Promise<T> newPromise(EventExecutor executor) {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/Promise.java:32]: Dunno if we actually want to do this. It doesn't seem very compelling, since there'd be few callers to it. (Most callsites will get promises from the channel.)

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 5]
```diff
@@ -68,17 +72,34 @@
...
      */
     public DefaultPromise(EventExecutor executor) {
         this.executor = requireNonNull(executor, "executor");
         stage = new DefaultFutureCompletionStage<>(this);
     }
 
+    /**
+     * Creates a completed promise.
+     */
+    DefaultPromise(EventExecutor executor, Object result) {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:88]: I like getting rid of FailedFuture and SucceededFuture. There's no real benefit to the specialized classes and they may prevent some JIT optimizations.

File: common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java (hunk 2 above)
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java:38]: Yeah, that simplifies the code a lot.

File: common/src/main/java/io/netty/util/concurrent/Promise.java [hunk 6]
```diff
@@ -57,9 +59,55 @@
...
+    /**
+     * Adds the specified listener to this future.
+     * The specified listener is notified when this future is {@linkplain #isDone() done}.
+     * If this future is already completed, the specified listener is notified immediately.
+     *
+     * @param listener The listener to be called when this future completes.
+     *                 The listener will be passed this future as an argument.
+     * @return this future object.
+     */
+    Future<V> addListener(FutureListener<? super V> listener);
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/Promise.java:100]: Why is this on Promise? Seems this just encourages the asking-for-bugs approach where code adds a listener to the promise before passing it further down the pipeline, instead of adding it to the returned future.

I'm fine if we just don't want to make that change in this PR, but if we are changing Promise this substantially, then we should fix that gap as well.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 7]
```diff
@@ -63,22 +68,51 @@
...
+    /**
+     * Creates a new promise that has already failed.
+     *
+     * @param executor
+     *        The {@link EventExecutor} which is used to notify the promise once it is complete.
+     *        It is assumed this executor will protect against {@link StackOverflowError} exceptions.
+     *        The executor may be used to avoid {@link StackOverflowError} by executing a {@link Runnable} if the stack
+     *        depth exceeds a threshold.
+     */
+    public DefaultPromise(Throwable cause, EventExecutor executor) {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:110]: This argument ordering difference seems to just be asking for bugs with code calling wrong constructor. These should probably be static factory methods in DefaultPromise for the failed/completed cases so that the `EventExecutor` is always first and the method name makes it clear what you are constructing.

File: common/src/main/java/io/netty/util/concurrent/Promise.java [hunk 8]
```diff
@@ -15,67 +15,99 @@
  */
 package io.netty.util.concurrent;
 
+import java.util.concurrent.CancellationException;
+
 /**
  * Special {@link Future} which is writable.
  */
-public interface Promise<V> extends Future<V> {
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/Promise.java:21]: Can we not bite this off in this PR? It seemed like the PR is changing enough as is, and this seems fundamentally a separate change.

It isn't clear to me how much value this adds, but I would _really_ appreciate that it'd make it harder to add listeners to promises which causes weird listener orderings. I don't _dislike_ it, but I have trouble identifying a way it provides enough value to justify the API migration.

**Review by ejona86 — COMMENTED** [review]
I feel comfortable with the FutureListener/Context stuff. And I see that it'd be easy to backport some of the changes to 4.x to allow absorbing this API change before jumping to netty 5.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java (hunk 7 above)
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:110]: Static methods on classes show up on instances, though. That's why I initially put those static methods on the interfaces.

File: common/src/main/java/io/netty/util/concurrent/Promise.java (hunk 6 above)
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/Promise.java:100]: Use cases like that showed up a lot, so it seemed convenient. Maybe I'll revisit later.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java (hunk 7 above)
Comment (ejona86) [reply common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:110]: Static method on DefaultPromise seems strongly better than this easy-to-use-wrong-constructor. The chance of calling `existingPromise.newFailedPromise()` seems rather low, especially since many callsites will use Promise-typed variables, and if it happens it causes almost no harm. I believe there's multiple linters that notice issues like that.

File: common/src/main/java/io/netty/util/concurrent/Promise.java (hunk 6 above)
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/Promise.java:100]: Let's punt on this change for now, since we'll keep having Promise extend Future.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpChunkedInput.java [hunk 9]
```diff
@@ -36,7 +36,7 @@
  *
  *     HttpContentChunkedInput httpChunkWriter = new HttpChunkedInput(
  *         new ChunkedFile(&quot;/tmp/myfile.txt&quot;));
- *     ChannelFuture sendFileFuture = ctx.write(httpChunkWriter);
+ *     Future sendFileFuture = ctx.write(httpChunkWriter);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpChunkedInput.java:39]:
```suggestion
 *     Future&ltVoid&gt sendFileFuture = ctx.write(httpChunkWriter);
```

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java [hunk 10]
```diff
@@ -143,9 +142,9 @@ private void handlePotentialUpgrade(final ChannelHandlerContext ctx,
                     headers.set(HttpHeaderNames.SEC_WEBSOCKET_EXTENSIONS, newHeaderValue);
                 }
             }
-            promise.addListener((ChannelFutureListener) future -> {
+            promise.addListener(future -> {
                 if (future.isSuccess()) {
-                    ctx.pipeline().remove(WebSocketServerExtensionHandler.this);
+                    ctx.pipeline().remove(this);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java:147]: I am surprised this works :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java [hunk 11]
```diff
@@ -97,12 +96,12 @@ public Handle newHandle() {
     private static final AtomicIntegerFieldUpdater<AbstractHttp2StreamChannel> UNWRITABLE_UPDATER =
             AtomicIntegerFieldUpdater.newUpdater(AbstractHttp2StreamChannel.class, "unwritable");
 
-    private static void windowUpdateFrameWriteComplete(ChannelFuture future, Channel streamChannel) {
+    private static void windowUpdateFrameWriteComplete(Channel streamChannel, Future<?> future) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java:99]: nit: Why this change ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java [hunk 12]
```diff
@@ -1011,7 +1008,7 @@ public void flush() {
                 // There is nothing to flush so this is a NOOP.
                 return;
             }
-            // We need to set this to false before we call flush0(...) as ChannelFutureListener may produce more data
+            // We need to set this to false before we call flush0(...) as ChannelFutureListeners may produce more data
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java:1011]:
```suggestion
            // We need to set this to false before we call flush0(...) as FutureListener may produce more data
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 13]
```diff
@@ -623,10 +619,4 @@ private static void verifyErrorCode(long errorCode) {
     private static void verifyWindowSizeIncrement(int windowSizeIncrement) {
         checkPositiveOrZero(windowSizeIncrement, "windowSizeIncrement");
     }
-
-    private static void verifyPingPayload(ByteBuf data) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:627]: should this change be something that is done outside of this PR ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java [hunk 14]
```diff
@@ -91,10 +90,10 @@
  *
  * <pre>{@code
  *     final Http2Stream2 stream = handler.newStream();
- *     ctx.write(headersFrame.stream(stream)).addListener(new ChannelFutureListener() {
+ *     ctx.write(headersFrame.stream(stream)).addListener(new FutureListener() {
  *
  *         @Override
- *         public void operationComplete(ChannelFuture f) {
+ *         public void operationComplete(Future f) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:96]: I think the `<Void>`is missing here. Also the same is true for `FutureListener` above.

**Review by normanmaurer — CHANGES_REQUESTED** [review]
All in all I really like the change! Great work!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java (hunk 11 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java:99]: It's used as a method reference, and therefor the signature needs to be compatible with the `FutureContextListener` interface.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java (hunk 10 above)
Comment (chrisvest) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java:147]: Lambdas don't have their own `this`, so now that this is no longer an anonymous inner class, we don't have to qualify the `this` reference.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 13 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:627]: Some unused code I came across. Want me to make a separate PR for it?

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:627]: Yep as I am almost sure this is also present in 4.1

File: common/src/main/java/io/netty/util/concurrent/Future.java [hunk 15]
```diff
@@ -49,12 +48,27 @@
...
+     * The specified listener is notified when this future is {@linkplain #isDone() done}.
+     * If this future is already completed, the specified listener is notified immediately.
+     *
+     * @param context The context object that will be passed to the listener when this future completes.
+     * @param listener The listener to be called when this future completes.
+     *                 The listener will be passed the given context, and this future.
+     * @return this future object.
      */
-    Future<V> addListener(GenericFutureListener<? extends Future<? super V>> listener);
+    <C> Future<V> addListener(C context, FutureContextListener<? super C, ? super V> listener);
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/Future.java:71]: I wonder if we really need this. Why not add something to FutureListener which allows to create a FutureListener that does the same.

Something like:

```
interface FutureListener {

    static <C, V> FutureListener newContextAwareListener(C context, BiConsumer<C, V> consumer);
}
```

Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/Future.java:71]: That approach will require more allocation. By passing the context in separately when adding the listener, the lambdas don't need to be capturing for the purpose of the context object at least.

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/Future.java:71]: I see... I was just wondering if we should keep things "slim" in terms of API surface. Not strong on this tho.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java (hunk 10 above)
Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java:147]: TIL!

**Review by NiteshKant — CHANGES_REQUESTED** [review]
Overall this looks great, few comments and I am still on the fence about the additional API but it seems like a lesser evil than the future/promise hierarchy

File: common/src/main/java/io/netty/util/concurrent/FutureListener.java [hunk 16]
```diff
@@ -16,13 +16,21 @@
...
  * <pre>
  * Future f = new DefaultPromise(..);
- * f.addListener(new FutureListener() {
- *     public void operationComplete(Future f) { .. }
- * });
+ * f.addListener(future -> { .. });
  * </pre>
  */
-public interface FutureListener<V> extends GenericFutureListener<Future<V>> { }
+public interface FutureListener<V> extends EventListener {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/FutureListener.java:29]: I don't know.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketProtocolHandler.java [hunk 17]
```diff
@@ -96,8 +96,8 @@ public void close(final ChannelHandlerContext ctx, final ChannelPromise promise)
     }
 
     @Override
-    public void write(final ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        if (closeSent != null) {
+    public void write(final ChannelHandlerContext ctx, Object msg, Promise<Void> promise) {
+        if (closeSent != null && closeSent != promise) {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketProtocolHandler.java:100]: There's some reentrancy that I'm not sure how was allowed to work before. Basically, when you close a web socket stream, it first assigns the closedSent promise, and writes the close frame, which gets us back into this method where we check if the stream has been closed. This normally prevents writes, but we can't allow ourselves to prevent the write of the close frame.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketServerHandshaker.java [hunk 18]
```diff
@@ -384,15 +383,15 @@ public ChannelFuture close(ChannelHandlerContext ctx, CloseWebSocketFrame frame)
...
      *            Closing Frame that was received.
      * @param promise
-     *            the {@link ChannelPromise} to be notified when the closing handshake is done.
+     *            the {@link Promise} to be notified when the closing handshake is done.
      */
-    public ChannelFuture close(ChannelHandlerContext ctx, CloseWebSocketFrame frame, ChannelPromise promise) {
+    public Future<Void> close(ChannelHandlerContext ctx, CloseWebSocketFrame frame, Promise<Void> promise) {
         requireNonNull(ctx, "ctx");
-        return close0(ctx, frame, promise).addListener(ChannelFutureListener.CLOSE);
+        return close0(ctx.channel(), frame, promise);
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketServerHandshaker.java:390]: Maybe that's the reason for the extra closedSent check above 🤔

File: common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java [hunk 19]
```diff
@@ -15,53 +15,77 @@
...
 
+import io.netty.util.internal.logging.InternalLogger;
+
 import java.util.Arrays;
+import java.util.EventListener;
 
 final class DefaultFutureListeners {
-
-    private GenericFutureListener<? extends Future<?>>[] listeners;
+    private Object[] listeners;
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java:24]: Listeners need to be called in order. Also, I've been trying to keep the allocations to a minimum.

File: transport/src/main/java/io/netty/channel/ChannelFuture.java [hunk 20]
```diff
@@ -1,187 +0,0 @@
...
-import io.netty.bootstrap.Bootstrap;
-import io.netty.util.concurrent.BlockingOperationException;
-import io.netty.util.concurrent.Future;
-import io.netty.util.concurrent.GenericFutureListener;
-
-import java.util.concurrent.TimeUnit;
-
-
-/**
- * The result of an asynchronous {@link Channel} I/O operation.
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelFuture.java:27]: I wonder if we want to move some of this javadoc... This looks like some useful documentation we don't want to loose.

File: transport/src/main/java/io/netty/channel/ChannelFutureListeners.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelFutureListeners.java:2]: 2021

File: transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java [hunk 22]
```diff
@@ -719,10 +723,12 @@ public ChannelHandlerContext flush() {
...
         } else {
             Tasks tasks = invokeTasks();
+            Promise<Void> promise = newPromise();
+            promise.addListener(channel(), ChannelFutureListeners.FIRE_EXCEPTION_ON_FAILURE);
             safeExecute(executor, tasks.invokeFlushTask,
-                    // If flush throws we want to at least propagate the exception through the ChannelPipeline
-                    // as otherwise the user will not be made aware of the failure at all.
-                    newPromise().addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE), null);
+                        // If flush throws we want to at least propagate the exception through the ChannelPipeline
+                        // as otherwise the user will not be made aware of the failure at all.
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java:730]: This comment should go above the `addListener` call

File: transport/src/test/java/io/netty/channel/DefaultChannelPipelineTest.java [hunk 23]
```diff
@@ -735,21 +730,22 @@ public void testWrongPromiseChannel() throws Exception {
         pipeline2.channel().register().sync();
 
         try {
-            ChannelPromise promise2 = pipeline2.channel().newPromise();
-            assertThrows(IllegalArgumentException.class, () -> pipeline.close(promise2));
+            Promise<Void> promise2 = pipeline2.channel().newPromise();
+            // Promises can complete across pipelines.
+            pipeline.close(promise2);
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/DefaultChannelPipelineTest.java:735]: 👍

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Great work! Once @NiteshKant and my comments are addressed this looks good to me! Need a strong ☕ now :)

File: common/src/test/java/io/netty/util/concurrent/DefaultPromiseTest.java [hunk 24]
```diff
@@ -71,60 +72,121 @@ private static int stackOverflowTestDepth() {
...
+        public void execute(Runnable command) {
+            fail("Cannot schedule commands");
+        }
+    }
+
     @Test
     public void testCancelDoesNotScheduleWhenNoListeners() {
-        EventExecutor executor = Mockito.mock(EventExecutor.class);
-        Mockito.when(executor.inEventLoop()).thenReturn(false);
+        EventExecutor executor = new RejectingEventExecutor();
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply common/src/test/java/io/netty/util/concurrent/DefaultPromiseTest.java:144]: General change that can be extracted.

[the remark this answers is not in the document]
Comment (chrisvest) [reply common/src/test/java/io/netty/util/concurrent/DefaultPromiseTest.java:144]: https://github.com/netty/netty/pull/11600

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ClientUpgradeCodecTest.java [hunk 25]
```diff
@@ -71,7 +71,10 @@ private static void testUpgrade(Http2ConnectionHandler handler, Http2MultiplexHa
         // Flush the channel to ensure we write out all buffered data
         channel.flush();
 
-        codec.upgradeTo(ctx, null);
+        channel.eventLoop().submit(() -> {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ClientUpgradeCodecTest.java:74]: No, it turns out `EmbeddedEventLoop` works differently between 4.1 and 5.x, so that these changes should not be back ported.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java [hunk 26]
```diff
@@ -618,30 +614,30 @@ public void register(ChannelPromise promise) {
...
             }
             if (closeInitiated) {
                 if (closePromise.isDone()) {
                     // Closed already.
-                    promise.setSuccess();
+                    promise.setSuccess(null);
                 } else  {
                     // This means close() was called before so we just register a listener and return
-                    closePromise.addListener((ChannelFutureListener) future -> promise.setSuccess());
+                    closePromise.addListener(promise, (p, future) -> p.setSuccess(null));
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java:640]: If the promise already has room for the listener, yeah. Non-capturing lambdas turn into a constant that gets reused.
If the promise has no listeners, then we end up allocating a `DefaultFutureListeners`, and an array. If that already exists, we could end up expanding that array.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamChannelBootstrap.java [hunk 27]
```diff
@@ -188,8 +183,8 @@ public void open0(ChannelHandlerContext ctx, final Promise<Http2StreamChannel> p
             return;
         }
 
-        ChannelFuture future = streamChannel.register();
-        future.addListener((ChannelFutureListener) future1 -> {
+        Future<Void> future = streamChannel.register();
+        future.addListener(future1 -> {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamChannelBootstrap.java:188]: The `promise` variable is also captured, so it doesn't win us anything.

File: common/src/main/java/io/netty/util/concurrent/CompleteFuture.java [hunk 28]
```diff
@@ -23,6 +23,7 @@
 /**
  * A skeletal {@link Future} implementation which represents a {@link Future} which has been completed already.
  */
+@Deprecated(forRemoval = true)
 public abstract class CompleteFuture<V> implements Future<V> {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/CompleteFuture.java:27]: yeah, missed that one

File: transport/src/main/java/io/netty/channel/ChannelFuture.java (hunk 20 above)
Comment (chrisvest) [reply transport/src/main/java/io/netty/channel/ChannelFuture.java:27]: Copied it over. Not sure if the class links work still.

**Comment by chrisvest** [comment]
@NiteshKant @normanmaurer Comments have been addressed.

**Review by normanmaurer — APPROVED** [review]

File: common/src/main/java/io/netty/util/concurrent/FutureListener.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply common/src/main/java/io/netty/util/concurrent/FutureListener.java:30]: I think we should remove it

File: common/src/main/java/io/netty/util/concurrent/Future.java (hunk 15 above)
Comment (NiteshKant) [reply common/src/main/java/io/netty/util/concurrent/Future.java:212]: @chrisvest any thoughts about splitting the API?

**Review by NiteshKant — APPROVED** [review]
Great work! Couple open discussions can be addressed as follow up if required.

File: common/src/main/java/io/netty/util/concurrent/Future.java (hunk 15 above)
Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/Future.java:212]: @NiteshKant Not a fan of splitting it. For one, I don't think the API is very complicated as is, and second, it found a lot of immediate usages when introduced here. I also briefly looked into #8523 today, and realised it can be used as a vehicle for future composition as well. Splitting the API as you've shown adds more moving parts, and adds a risk that some usage sites might once again care about the type of future, which was the reason we had `GenericFutureListener` and one of the main pain points this PR seeks to address. Not saying it's as bad with the `from()`, but it's a less clean break, IMHO.

Comment (NiteshKant) [reply common/src/main/java/io/netty/util/concurrent/Future.java:212]: Fair enough, thanks for sharing your views!

**Comment by chrisvest** [comment]
I'm gonna go ahead and merge this. @trustin if you have review comments, I'll be happy to address them in follow-up PRs.

**Comment by trustin** [comment]
Nice work, @chrisvest. Be gone, technical debts! :heart:
